File: Adhiti_Shaili/Day 23,24/Copy list with random pointer.cpp
#include<bits/stdc++.h>
using namespace std;
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:

    Node* copyRandomList(Node* head) {
        
        if(head==NULL)
        {
            return NULL;
        }
       
        Node* curr=head;
        while(curr!=NULL)
        {
            Node* temp=curr->next;
            Node* currRep=new Node(curr->val);
            curr->next=currRep;
            currRep->next=temp;
            curr=curr->next->next;
        }
        curr=head;
        while(curr!=NULL)
        {
            if(curr->random!=NULL)
            {
                curr->next->random=curr->random->next;
            }
            else
            {
                curr->next->random=NULL;
            }
            curr=curr->next->next;
        }
        
        curr=head;
        Node* newHead=curr->next;
        Node* newCurr=newHead;
        while(curr->next!=NULL&&newCurr->next!=NULL)
        {
            curr->next=curr->next->next;
            curr=curr->next;
            newCurr->next=newCurr->next->next;
            newCurr=newCurr->next;
        }
        curr->next=NULL;
        return newHead;
    }
};

File: parnamondal/Week - 5/Day - 2/Copy List with Random Pointer.py
# Time Complexity- O(N).
# Space Complexity - O(1)

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        temp=head
        def insert(head):
            if head==None:
                return None
            cur=head
            r=Node(cur.val)
            later=cur.next
            cur.next=r
            r.next=later
            insert(later)
        def rand(head):
            if head==None:
                return None
            t=head
            if t.random!=None:
                temp=t.random
                t.next.random=temp.next
            rand(t.next.next)
        def getlist(curr):
            if curr==None:
                return
            ret=curr.next
            curr.next=ret.next
            ret.next=getlist(ret.next)
            return ret
        insert(temp)
        rand(temp)
        t=getlist(temp)
        return t
    
        